### DISCUSSION TITLE
What exactly does the stable URI represent?

### DISCUSSION THREAD

**Discussion by jdesrosiers** [json-schema-org/community#438] [Specification]
In previous discussion, I though we had agreed on what the stable dialect URI would represent. My understanding was that there was agreement on what I describe [here](https://github.com/orgs/json-schema-org/discussions/384#discussioncomment-5774354), but it seems there's still some [disagreement](https://github.com/json-schema-org/json-schema-spec/issues/1421#issuecomment-1622587099), so let's try again.

Please read the linked comments, but discuss here.

**Comment by jdesrosiers** [comment]
IMO, allowing for the stable URI to point to an implementation-specific meta-schema rather than being limited to official released meta-schemas is critical to the processes we've been discussing for the staged introduction of new keywords. New keywords can only reach "stable" status if they are used. The sooner those keywords can be made available to users, the sooner they can reach "stable" status, which is good for everyone.

What I have in mind is similar to how JavaScript evolves. Implementations start adding support for new features as soon as they're proposed and people start using them as soon as their implemented. There's no lag between spec release and implementations supporting the feature because they already support the feature. In fact, the implementations supporting the feature was a prerequisite for making it into the spec release.

I should be able to make a keyword available as soon as it's proposed and not have to wait as much as a year or more before a new meta-schema is released so people can start using it. I should be able to update my implementation specific meta-schema to include the new keyword without having to wait for the official yearly release. An alternative to this problem is for us to make a living meta-schema available to implementers that includes proposed keywords as soon as they're proposed, but that has some problems I won't go into right now unless someone asks to hear more.

A second problem with limiting to release meta-schemas is accounting for the unique needs of some implementations. For example, I need an open meta-schema for validating schemas. Validating schemas is optional, so I can't rely on `unevaluatedProperties` to check if a keyword is supported or not. Therefore, I explicitly check each keyword that is encountered for support and provide a much nicer error message that I can get with `unevaluatedProperties`. I want users to always get the nice message from the additional check rather than sometimes get a different `unealuatedProperties` error.

In order to get that behavior, I would have to alter the official meta-schema to remove the `unevalutedProperties` to get the best experience for my users. If the spec forbids such alterations, I'm going to do it anyway because it's good for my users and doesn't hurt anything. No one would even know the difference unless they read my code. The outcomes are always the same even tho I'm technically violating the spec.

I think there are some constraints we can and should put on what can be in an implementation-specific meta-schema that the stable URI is allowed to point to, but limiting to the release meta-schemas is too restrictive. An implementation-specific meta-schema must be compatible with a release meta-schema with the exception of `unevaluatedProperties`. That means you can add keywords, but you can't remove or modify keywords or modify `$vocabularies`. I think that's reasonable, but like in the previous example, this restriction is completely unenforceable and implementations are going to ignore it if it suits their needs.

**Reply by gregsdennis** [reply]
> An alternative to this problem is for us to make a living meta-schema available to implementers that includes proposed keywords as soon as they're proposed, but that has some problems I won't go into right now unless someone asks to hear more.

This is precisely what I thought we were doing, so, yeah, I'd like to hear what problems you think this has.

As I understand it, the tentative process that we're going for is that keywords can be added to the spec as proposals (stage 1).  I would expect that the meta-schema is updated as part of that.  This keeps the spec and meta-schema in alignment.  It allows us to verify how the meta-schema is going to change (rather than trying to define it just before we take a snapshot).  And it allows the closed version to adopt the new keyword.

Snapshots of the spec and meta-schema will continue to be taken annually, as planned.  But I see no issues with keeping the meta-schema aligned with the spec.

Conversely, I see a problems with _not_ updating the meta-schema when the spec changes.  Specifically, (what you pointed out) the closed meta-schema will reject new/altered keywords when the spec defines/alters/redefines a stage 1 keyword (as will happen because stage 1 isn't stable).  Not ideal if we want people to use the new keyword.

A live meta-schema is the only way that I can see to ensure keywords are available as soon as they're proposed.  The meta-schema and spec must remain in sync.  Adding a keyword to the meta-schema needs to be part of the stage 1 requirements.

> A second problem with limiting to release meta-schemas is accounting for the unique needs of some implementations. For example, I need an open meta-schema for validating schemas. Validating schemas is optional, so I can't rely on unevaluatedProperties to check if a keyword is supported or not. Therefore, I explicitly check each keyword that is encountered for support and provide a much nicer error message that I can get with unevaluatedProperties. I want users to always get the nice message from the additional check rather than sometimes get a different unealuatedProperties error.
>
> In order to get that behavior, I would have to alter the official meta-schema to remove the unevalutedProperties to get the best experience for my users. If the spec forbids such alterations, I'm going to do it anyway because it's good for my users and doesn't hurt anything. No one would even know the difference unless they read my code. The outcomes are always the same even tho I'm technically violating the spec.

All of this is moot if we:

- keep a live meta-schema that's aligned with the live spec
- keep an open and a closed version of the live meta-schema

If we do these two things, you get your cake and the ability to eat it.

## Implementation Support

Let's consider the scenarios.  There are two sets: what an implementation supports and whether the implementation packages the meta-schema.

1. An implementation fully supports everything that was in the latest specification when the implementation package was published.
2. An implementation supports everything in a snapshot plus some set of features introduced since that snapshot.
3. An implementation supports everything in a snapshot and nothing more.
4. An implementation only supports bits and pieces, not fully supporting any snapshot.

A. An implementation packages the meta-schema.  From the user's perspective, the meta-schema cannot change apart from the implementation releasing a new package.
B. An implementation uses the online or user-provided meta-schema.  The meta-schema _can_ change without the implementation releasing a new package.

(markdown formatting isn't the greatest sometimes)

### 1. Full support of bleeding edge of JSON Schema

If the implementation packages the meta-schema, then it should package the live meta-schema in its current state.  Schema validation always works and is perfectly aligned with implementation support.

If the implementation uses the online or user-provided meta-schema, then

- it should be expected that future changes to the live meta-schema could result in incompatible changes to unstable keywords.  Thus the implementation needs to be updated.  The user either finds a newer version of the implementation or files an issue.
- it should be expected that future changes to the live meta-schema could introduce new keywords.  The implementation doesn't support these, so it will error if it encounters a schema that uses them.  The user either finds a newer version of the implementation or files an issue.

### 2. Full support of a snapshot plus select newer features

If the implementation packages the meta-schema, then it should package the live meta-schema in its current state.  It should be expected that the meta-schema may declare features not supported by the implementation, but the implementation will error if any of those are encountered.  The user either finds a newer version of the implementation or files an issue.

If the implementation uses the online or user-provided meta-schema, we have the same results as (1).

### 3. Full support of a snapshot and nothing more.

If the implementation packages the meta-schema, then it should package the snapshot associated with its support.  Schema validation always works and is perfectly aligned with the implementation.

If the implementation uses the online or user-provided meta-schema, it should be expected that a later meta-schema could introduce new keywords.  The implementation doesn't support these, so it will error if it encounters a schema that uses them.  The user either finds a newer version of the implementation or files an issue.

### 4. Partial support of everything.

If the implementation packages the meta-schema, then it should package the earliest meta-schema that covers the implementation's support.  It should be expected that the meta-schema may (and likely does) declare features not supported by the implementation, but the implementation will error if any of those are encountered.  The user either finds a newer version of the implementation or files an issue.

If the implementation uses the online or user-provided meta-schema, it should be expected that a later meta-schema could (and likely will) introduce new keywords.  The implementation doesn't support these, so it will error if it encounters a schema that uses them.  The user either finds a newer version of the implementation or files an issue.

(I sincerely hope we don't have any implementations in this bucket.)

### Summary

None of these cases warrant using a meta-schema that we (JSON Schema) didn't publish.  Implementations should always be able to use unmodified official meta-schemas.

But this only works if we have a live meta-schema that is kept in sync with the specification.

## Caveats

The only cases that I can see for modifying the meta-schema is if you're defining a new keyword in your implementation

- in order to get it over that stage 1 hurdle (requires two implementations).  In this case, it's experimental, and the addition of the new keyword is part of the proposal.
- because you can, but have no intention of proposing it (or your proposal was rejected so you did it anyway).  In this case, the implementation is operating outside the scope of the spec.

Since these are outside of the normal expectations of an implementation, I don't think they're relevant to the discussion.

**Reply by jdesrosiers** [reply]
> The only cases that I can see for modifying the meta-schema is if you're defining a new keyword in your implementation
>
> * in order to get it over that stage 1 hurdle (requires two implementations). In this case, it's experimental, and the addition of the new keyword is part of the proposal.
> * [...]
>
> Since these are outside of the normal expectations of an implementation, I don't think they're relevant to the discussion.

How is this not relevant? If the spec says an implementation can only use official meta-schemas, proposals can't be implemented without violating the spec. It seems like a significant problem that the first step in the process of introducing a new keyword requires violating the spec. The alternate solution I was suggesting was about maintaining a meta-schema that includes proposals (stage-0, not yet in the spec).

What you're describing is maintaining a meta-schema that matches the spec as it evolves (not including stage-0). I fully expect for us to continue doing that. What I wasn't expecting was that we would publish that schema or that you considered it part of the set of meta-schemas implementations were allowed to use. Doing that would addresses the major issues I have with being more strict (with the exception of supporting stage-0), but I think there are still some cases where slight customization would be necessary.

Imagine there are two stage-1 keywords in the spec: `foo` and `bar`. A change to `foo` is needed which requires an incompatible meta-schema change. The change is not easy to implement and I (the implementer) don't have the capacity to make that change just yet. Shortly after a change to `bar` is needed which also requires an incompatible meta-schema change. The `bar` change is simple and I can push it out in a few minutes, but I can't do that without also making the harder changes to `foo` because there is no official meta-schema that includes the new version of `bar` and the old version of `foo`. My choice is to follow the letter of the spec and hold back the progress of `bar`, or violate the spec and customize my meta-schema.

Keywords are going to progress through the stabilization process at different rates and we need to allow for the case where implementers aren't necessarily updating their implementations in same order that we update the spec. In theory, a keyword should be stable enough by the time it reaches stage-1 that a scenario like the one above occurring should be an extreme statistical outlier, but it is possible, so I think think we should account for it.

However, if we solve for the stage-0 problem with a published living meta-schema that includes proposals, the scenario I describe could be a common one because those keywords are much more likely to be in flux and not implemented in order.

**Reply by gregsdennis** [reply]
> If the spec says an implementation can only use official meta-schemas...

Please allow me to clarify.  What I mean is that if a packaged meta-schema has a URI that matches an official meta-schema URI, then the meta-schema itself should match what we publish.  What I don't want is for someone to choose not to support (e.g.) `minimum` and then to publish their package with a meta-schema that uses our URI but excludes the `minimum` keyword.  The way you stated allowing implementations to package a modified meta-schema would allow this.

Proposals are a special case because (1) they're experimental and (2) they're additions.  They're a stepping stone to including a feature in the spec.  I agree that we should include exceptions/provisions for proposals, specifically.   I don't have a problem with this, but we need to be clear that it only pertains to proposals, not stable features and not non-proposal additions (e.g. `$data`).

Those portions of the meta-schema which represent stable (and deprecated) features need to be present as they appear in the meta-schema we publish.

**Reply by jdesrosiers** [reply]
> What I mean is that if a packaged meta-schema has a URI that matches an official meta-schema URI, then the meta-schema itself should match what we publish.

Agreed. If an implementation uses a custom meta-schema, it should have a custom identifier. My understanding is that the point in contention is whether or not the stable dialect URI is allowed to point to that custom meta-schema. Based on this latest comment, it sounds like you're ok with that, but with strict requirements on what that custom meta-schema is allowed to contain.

> The way you stated allowing implementations to package a modified meta-schema would allow this.

My wording may have been misleading. I'm never sure how precisely I need to say things and when a shorthand will be understood. When I talked about a "modified meta-schema", I meant a copy with necessary modifications. It would have a unique identifier from the meta-schema it was copied/altered from. However, this meta-schema's only purpose is to be used internally as the target of the stable dialect URI, so users would never use that identifier explicitly in `$schema` or anywhere else.

(Since you rely on static typing to validate schemas, you would never have need for an internal implementation-specific meta-schema. Your types serve the purpose that the meta-schema serves for me.)

> I agree that we should include exceptions/provisions for proposals, specifically. I don't have a problem with this, but we need to be clear that it only pertains to proposals, not stable features and not non-proposal additions (e.g. `$data`).
>
> Those portions of the meta-schema which represent stable (and deprecated) features need to be present as they appear in the meta-schema we publish.

Agreed. The stable URI should only be allowed to point to a meta-schema that is compatible with the standard dialect and official proposals. If an implementation wants to support a keyword that is not accepted as an official proposal, they would have to support it as a custom vocabulary that people would have to use a custom dialect to enable and use.

**Reply by gregsdennis** [reply]
> However, this meta-schema's only purpose is to be used internally... so users would never use that identifier explicitly in $schema or anywhere else.

If this sort of custom meta-schema is only supposed to be used internally, why are we bothering with specifying it?  Sounds like it just became an implementation detail to me.  If this is something that your implementation needs in order to operate and be compliant, and it isn't exposed to your users, I have no problem with it, and I don't think the spec has the authority to even comment about this practice.

**Reply by jdesrosiers** [reply]
Yes! Now we're really getting on the same page. The spec just needs to define that the stable dialect URI represents the most recent version of JSON Schema the implementation supports, potentially including accepted proposals not yet in the spec. We don't need to limit what meta-schema an implementation associates with that URI (if any). It's up to the implementation to use whatever is necessary to operate as is expected when the stable dialect URI is used.

**Reply by gregsdennis** [reply]
Does this mean you're on board with a living meta-schema as well?  It would stay in sync with the spec and be updated as proposals reach stage 1.  Moreover, the changes to the meta-schema would be _part of_ stage 1 requirements.

This assumes that the implementor retains the right to maintain an internal version of the meta-schema that is only modified for proposals in order to allow the implementor to update support for those proposals as they can.  Stable features MUST be included as they are in the official meta-schema.

(This is going to be fun to enforce.)

**Reply by jdesrosiers** [reply]
Yes. :100: to all of that.

**Reply by Relequestual** [reply]
I think I agree with all of the above discussion, however something occured to me.

If an implementation modifies the meta-schema, and reports an error based on the modified meta-schema...

1. What base URI will it report? The official one, that they used, or a custom one the implementation defined?
2. How will a user understand a URI path into the meta-schema which seems to not exist in the officially published meta-schema?

Expanding on 2, sure, you could argue that the user should know they are using experimental non-stable features, but I can also imagine that someone leaves and someone else has to pick up a bunch of schemas without much of a clue, tries to modify the schema, and gets an error they don't understand how to digest, because they can't find the URI path of the meta-schema.

Thoughts?

**Reply by gregsdennis** [reply]
First, remember that this is _meta-schema_ evaluation, which doesn't really happen a lot as it's optional.  The primary cases for validating with the meta-schema is when you're authoring a schema.  (And then, it's likely the editor that's using it, not the author directly.)

1. The URI will still be the stable URI.  Stable features will still be well-defined.  The proposals that the implementation supports will need to be documented.
2. Per [this thread](https://github.com/orgs/json-schema-org/discussions/435#discussioncomment-6443964), we're looking at not hosting the stable URI meta-schema, so it's not "officially published."  The meta-schema is live, includes all current proposals, and is _available_ on GitHub, but it's not published.
2+.  I would hope that a dev picking up work would go to the implementation's documentation.  It would probably be a good idea for us to publish documentation about expectations for support of proposals and other unstable features that's not specific to any implementation.
